myco_api_client/matchers/redis_base_matcher.py

```diff
@@ -97,5 +97,14 @@ def _on_event_or_response(self, payload: Dict):
         self.executor.submit(self.on_event_or_response, data)
         # Call the corresponding event handler
         event = data.get("event")
-        if hasattr(self, f"_on_{event}"):
-            self.executor.submit(getattr(self, f"_on_{event}"), data)
+        callback_function_name = f"_on_{event}"
+        if hasattr(self, callback_function_name):
+            # Schedule the callback
+            future = self.executor.submit(getattr(self, callback_function_name), data)
+            try:
+                # Explicitly execute the call. Not doing so will cause the ThreadPoolExecutor to
+                # swallow and hide potential exceptions making it tricky to debug.
+                future.result()
```

Comment:
AFAIU from [here](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result),  this will block until the `getattr(self, f"_on_{event}")` has returned. I dont know if this will be tricky for the myco client who is listening to the `_on_event_or_response` in order to react on it. What do you think?

Reply:
TL;DR: I think you're right. If we use `future.result()` on each call, then the next call will need to wait until the previous one has finished. This problem was already experienced and fixed in the `d3a-api-client`. I'll use the same strategy, but this will involve moving one util to d3a-interface: https://github.com/gridsingularity/d3a-api-client/blob/d66f962e7aede90de29ef5eeeaaadbebd0d32c9c/d3a_api_client/utils.py#L174-L179)

---

Longer recap for my own clarity:
1. The executor is used by `RedisMycoMatcher(RedisBaseMatcher)` instances. The way it's used is the following: https://github.com/gridsingularity/myco-api-client/blob/8f31f92c5ae772211d3fddccea98ef5a7ab4e0f1/myco_api_client/setups/redis_myco_matcher.py#L40-L43
So there will be only one instance of this matcher (and one instance of `matcher.executor`) that survives during the entire simulation.
2. Each time the matcher receives an event, the executor submits the relative call and creates a thread for it.
3. We can't use `.result()` on each future, because this would block the execution of the next one. Since we don't have a list of all the current scheduled futures, we can't use `concurrent.futures.as_completed(futures)` (which is the way to make the operations non-blocking).
4. In order to fix this, we can wrap each callback in a try/except block and log the errors there.